edit-post/hooks/autocompleters/index.js

```diff
@@ -0,0 +1,44 @@
+/**
+ * WordPress dependencies.
+ */
+import { addFilter } from '@wordpress/hooks';
+import { select } from '@wordpress/data';
+import { createBlock } from '@wordpress/blocks';
+
+// Export for unit test
+export function getBlockOptions( query ) {
+	const editor = select( 'core/editor' );
+	const isInitialQuery = ! query;
+
+	// TODO: Fetch reusable blocks if necessary (they aren't fetched initially)
+	return isInitialQuery ?
+		// Before we have a query, offer frecent blocks as a sensible default.
+		editor.getFrecentInserterItems() :
+		editor.getInserterItems();
```

Comment:
Will this list of items stay current when shared blocks are added and deleted? Just checking since usually we use `withSelect` which automatically subscribes us to store updates, but in this case we are calling `select` manually.

Also, we will need to pass the `enabledBlockTypes` argument to these selectors so that blocks that have been disabled via the `allowed_block_types` filter cannot be inserted into the post. Usually, [we get this setting by using `withContext`](https://github.com/WordPress/gutenberg/find/master#L343-L349), but I'm not sure how that translates to this since we're not working with a component.

Reply:
A completer provides a static list of options, but `Autocomplete` will request a new list for each keystroke. If a completer's `options` property is a function, `Autocomplete` will call it to get the next options, and those options can be based on updated data like this.

> Also, we will need to pass the `enabledBlockTypes` argument to these selectors so that blocks that have been disabled via the `allowed_block_types` filter cannot be inserted into the post.

I wondered about this but noticed that the default for `enabledBlockTypes` is true for those functions so we are getting the desired behavior by default. Am I missing something?

Reply:
> I wondered about this but noticed that the default for enabledBlockTypes is true for those functions so we are getting the desired behavior by default. Am I missing something?

Yes, the idea is that the caller of `getInserterItems()` should [pass along an array of allowed block types](https://github.com/WordPress/gutenberg/blob/master/editor/components/inserter/menu.js#L353-L354) so that plugin developers can turn specific block types on or off [via a filter](https://github.com/WordPress/gutenberg/blob/master/lib/client-assets.php#L934).

(I regret making the argument optional.)

Reply:
Thanks for explaining.

> Usually, [we get this setting by using `withContext`](https://github.com/WordPress/gutenberg/find/master#L343-L349), but I'm not sure how that translates to this since we're not working with a component.

This could be interesting, but it's good we're feeling this friction now. I'll see what I can find.

Reply:
It turns out that this issue exists in master, so feel free to ignore it for now if you want. I opened https://github.com/WordPress/gutenberg/issues/6070 to track it.